clientlibrary/checkpoint/dynamodb-leasestealer.go

```diff
@@ -0,0 +1,190 @@
+ * copies of the Software, and to permit persons to whom the Software is furnished to do
+ * so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all copies or substantial
+ * portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT
+ * NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+ * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+// The implementation is derived from https://github.com/patrobinson/gokini
+//
+// Copyright 2018 Patrick robinson
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+package checkpoint
+
+import (
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/client"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/dynamodb"
+	"github.com/aws/aws-sdk-go/service/dynamodb/dynamodbiface"
+
+	"github.com/vmware/vmware-go-kcl/clientlibrary/config"
+	par "github.com/vmware/vmware-go-kcl/clientlibrary/partition"
+	"github.com/vmware/vmware-go-kcl/logger"
+)
+
+// DynamoLeasestealer implements the Leasestealer interface using DynamoDB as a backend
+type DynamoLeasestealer struct {
+	log          logger.Logger
+	TableName    string
+	svc          dynamodbiface.DynamoDBAPI
+	kclConfig    *config.KinesisClientLibConfiguration
+	Retries      int
+	checkpointer Checkpointer
+}
+
+func NewDynamoLeasestealer(kclConfig *config.KinesisClientLibConfiguration, checkpointer Checkpointer) *DynamoLeasestealer {
+	leasestealer := &DynamoLeasestealer{
+		log:          kclConfig.Logger,
+		TableName:    kclConfig.TableName,
```

Comment:
The checkpointer and stealer should share the same dynamodb table. There is no need to create an instance for this.

Reply:
So the mean reason for this was so that we could add a `ClaimShard` method and be backwards compatible. If we were to add `ClaimShard` to the checkpointer interface that would be a breaking change.

Reply:
We need table name because of: https://github.com/connormckelvey/vmware-go-kcl/blob/293d7796ff5216d2f1a7102ea57351371004fe0c/clientlibrary/checkpoint/dynamodb-leasestealer.go#L178

Reply:
I think this change is quite awkward. golang need to be compiled from source. I don't think this will affect much. If we end up introducing incompatibility, we do have good reason for it. We did this previously, see:
https://github.com/vmware/vmware-go-kcl/tree/v-1.0

Reply:
When we first talked you requested that lease stealing functionality be backwards compatible and non-breaking. This achieves that. 

I'm imagining a case where a user has a custom checkpointer defined that implements the checkpointer interface. For example if a user uses redis for checkpointing instead of dynamodb. Adding another method to the Checkpointer interface would break the uses implementation and they would need to implement a no-op method to satisfy the interface.

Reply:
This is an example on how to inject a custom checkpointer. User can provide their own implementation, redis or RDS, etc.
https://github.com/vmware/vmware-go-kcl/blob/master/test/worker_custom_test.go#L58